### PR TITLE
Dibi type inference

### PR DESCRIPTION
**Pull request by staabm**
building on top of https://github.com/staabm/phpstan-dba/pull/465 - all credits to [jakubvojacek](https://github.com/jakubvojacek)

### CODE REVIEW

File: src/Extensions/DibiConnectionFetchDynamicReturnTypeExtension.php [hunk 1]
```diff
@@ -0,0 +1,200 @@
...
+            return TypeCombinator::union(...$fetchTypes);
+        }
+        if (1 === \count($fetchTypes)) {
+            return $fetchTypes[0];
+        }
+
+        return null;
+    }
+
+    private function reduceResultType(MethodReflection $methodReflection, Type $resultType): ?Type
```
Comment (staabm) [inline src/Extensions/DibiConnectionFetchDynamicReturnTypeExtension.php:118]: todo adjust to match dibi return types

**Comment by staabm** [comment]
@jakubvojacek please have a look. based on https://github.com/staabm/phpstan-dba/pull/465 I have removed anything, which I think is not required (e.g. the DibiReflector). this PR needs polishing in `DibiConnectionFetchDynamicReturnTypeExtension` to make it return proper types in `reduceResultType` but otherwise should be good to go.

could you finish it from this state?

**Comment by jakubvojacek** [comment]
on it, thanks :+1:

As tomorrow is a public holiday, not sure I will manage to finish it

**Comment by staabm** [comment]
Take your time

**Comment by staabm** [comment]
only thing left seem to be the pgsql expectations @jakubvojacek :)

File: tests/rules/data/syntax-error-in-dibi-prepared-statement.php [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+        $conn->fetchPairs('SELECT email FROM ada');
+        $conn->fetchPairs('SELECT email,adaid FROM ada');
+    }
+
+    public function testPlaceholders(\Dibi\Connection $conn)
+    {
+        $conn->fetchPairs('SELECT email FROM ada where email = %s');
+        $conn->fetchPairs('SELECT email FROM ada where email = ""', 1);
+        $conn->fetchPairs('SELECT email FROM ada where email = %s', 'email@github.com');
+    }
```
Comment (staabm) [inline tests/rules/data/syntax-error-in-dibi-prepared-statement.php:42]: we need these placeholder test-cases also for typeinference in `dibi.php`

**Comment by staabm** [comment]
great job - thanks again.

I was not sure about your twitter handle, but I spreaded some news here: https://twitter.com/markusstaab/status/1593220018837544969

**Comment by jakubvojacek** [comment]
thanks, im not on twitter :)

**Pull request merged**